sys/fs/nfsclient/nfs_clvfsops.c

```diff
@@ -1243,26 +1243,23 @@ nfs_mount(struct mount *mp)
 			error = EINVAL;
 			goto out;
 		}
-		error = copyin(__USER_CAP(args.fh, args.fhsize), nfh,
-		    args.fhsize);
+		error = copyin(args.fh, nfh, args.fhsize);
 		if (error != 0)
 			goto out;
-		error = copyinstr(__USER_CAP(args.hostname, MNAMELEN), hst,
-		    MNAMELEN - 1, &hstlen);
+		error = copyinstr(args.hostname, hst, MNAMELEN - 1, &hstlen);
 		if (error != 0)
 			goto out;
 		bzero(&hst[hstlen], MNAMELEN - hstlen);
-		args.hostname = hst;
+		args.hostname = (__cheri_tocap char * __capability)hst;
```

Comment:
What I've done in some other places is to turn the pointer field into an anonymous union in the kernel with a plain pointer in the kernel and the user pointer as a cap with a "user_" prefix, e.g. in 'struct ccb_scsiio'.  If you did the same here you would end up replacing the bogus __USER_CAP()'s with 'args.user_fh'.  Note that the NFS client might be one we might want FREEBSD64 shims for (as opposed to the NFS server).  I've used the NFS client ok with a hybrid world prior to this which was probably depending on this hack.

Reply:
Thanks for the suggestion, that's a much smaller diff.